### PR TITLE
Support mixed-language batches in `WhisperGenerationMixin`

### LINKED ISSUE #29685: Support mixed-language batches in `WhisperGenerationMixin`
### Feature request

It is currently not possible to mix multiple languages in a single batch when running [Whisper](https://huggingface.co/docs/transformers/en/model_doc/whisper). The `language` argument only accepts a single string (as opposed to a separate language for each batch item), and if no language is passed and multiple languages are detected, [transcription will fail](https://github.com/huggingface/transformers/blob/5011908e10d9592eeb634f4940e0bc130d3edc69/src/transformers/models/whisper/generation_whisper.py#L1170-L1173).

I propose to enable passing a list of languages (`language: Optional[Union[str, List[str]]]`) in a batched transcription situation, as well as removing the restriction related to language detection.

### Motivation

Not being able to transcribe multiple languages in a single batch is clearly a limitation, especially when relying on auto-detection, but also in scenarios where the language is known.

The [error message](https://github.com/huggingface/transformers/blob/5011908e10d9592eeb634f4940e0bc130d3edc69/src/transformers/models/whisper/generation_whisper.py#L1172) states that `It is currently not supported to transcribe to different languages in a single batch.`, implying that it could be supported at some point.

### Your contribution

I have implemented this and I'm planning to submit a PR.

### PR DESCRIPTION
**Pull request by cifkao**
# What does this PR do?

Fixes #29685

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

 @ylacombe @sanchit-gandhi

### CODE REVIEW

File: tests/models/whisper/test_modeling_whisper.py [hunk 1]
```diff
@@ -547,10 +547,16 @@ def test_generate_language(self):
...
+        # test language token
         model.generate(input_features, language="<|en|>")
         # test language name
         model.generate(input_features, language="English")
+        # test language code list
+        model.generate(input_features, language=["en"] * input_features.shape[0])
+        # test language token list
+        model.generate(input_features, language=["<|en|>"] * input_features.shape[0])
+        # test language name list
+        model.generate(input_features, language=["English"] * input_features.shape[0])
```
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:559]: This is great!

Could you also test:
1. On this same test, checking that passing a list of languages that is not the same length as input_features does indeed raise an error ?
2. On a slow test, that you get the same results generating two samples with two languages at once vs generating each sample with one language at once ?

File: src/transformers/models/whisper/generation_whisper.py [hunk 2]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
-            if language_token not in generation_config.lang_to_id:
+            if len(language) != batch_size:
                 raise ValueError(
-                    f"{language_token} is not supported by this specific model as it is not in the `generation_config.lang_to_id`."
-                    "(You should just add it to the generation config)"
+                    "When passing a list of languages, the length of the list must match the batch size. "
+                    f"Expected length of {batch_size}, but got {len(language)}"
                 )
+        else:
+            language = [language] * batch_size
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:1155]: This is a nit, but I wonder if you could keep a list of length 1 here and repeat `init_tokens` batch_size times at the end of the method instead, WDYT ?

**Review by ylacombe — COMMENTED** [review]
This looks great to me! Thanks for the great work here! 

I've left a few comments here and there. Could you also make sure that all slow tests still pass ?

Also requesting a review from @sanchit-gandhi as the changes are pretty heavy!

File: src/transformers/models/whisper/generation_whisper.py (hunk 2 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1155]: Sounds like a good idea to me.

File: tests/models/whisper/test_modeling_whisper.py (hunk 1 above)
Comment (cifkao) [reply tests/models/whisper/test_modeling_whisper.py:559]: Done!

**Comment by cifkao** [comment]
@ylacombe I ran the slow tests in `test_modeling_whisper`. I get some failing tests, but the same ones fail on `main` as well, with the same output:
```
========================================================================================================================== short test summary info ===========================================================================================================================
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_multi_batch - assert [' While Porashaggy sits there, a cooing dove. He has gone, and gone for good," answered Polychrom, who had managed to squeeze into the room beside the dragon, and had witnessed the occurrences with much interest. I have remained a prisoner only because I wi...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_multi_batch_prev_cond - assert [" While poor shaggy sits there, a cooling dove. He has gone and gone for good, answered polychrome, who had managed to squeeze into the room beside the dragon, and had witnessed the occurrences with much interest. I have remained a prisoner only because I w...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_prompt_ids - AssertionError: assert 'quilter' in 'because you were sleeping instead of conquering the lovely rose princess has become a fiddle without a bow while poor shaggy sits there a cooing dove'
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_single_batch - assert [" Because you were sleeping instead of conquering, the lovely rose princess has become a fiddle without a bow, all poor ashaggy sits there, accooing dove. He has gone and gone for good. Antered polychrome who had managed to squeeze into the room beside the ...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_single_batch_prev_cond - assert [" Because you were sleeping instead of conquering, the lovely rose princess has become a fiddle without a bow, all poor ashaggy sits there, accooing dove. He has gone and gone for good. Antered polychrome who had managed to squeeze into the room beside the ...
==================================================================================================== 5 failed, 292 passed, 106 skipped, 120 warnings in 249.92s (0:04:09) ====================================================================================================
```

File: src/transformers/models/whisper/generation_whisper.py [hunk 3]
```diff
@@ -1167,49 +1190,47 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
-            if len(init_tokens) > 1:
-                init_tokens[1] = lang_id
-            else:
-                init_tokens.append(lang_id)
-
-        if task is not None:
-            if task in TASK_IDS:
-                init_tokens.append(generation_config.task_to_id[generation_config.task])
-                task_id = generation_config.task_to_id[generation_config.task]
+            for i in range(len(init_tokens)):
```
Comment (naveen-corpusant) [inline src/transformers/models/whisper/generation_whisper.py:1194]: `init_tokens` is only 1 long in the case of `language=None` at this point (because when you do the deepcopy on init_tokens, it assumes that a list of languages has been passsed in). We also need to set language for a later check in the function, since the dim of language is used in a check

To fix this, we can do:
```
id_to_lang = {v: k for k, v in generation_config.lang_to_id.items()}
language = [id_to_lang[lang_id.item()] for lang_id in lang_ids]
init_tokens = [copy.deepcopy(init_tokens[0]) for _ in lang_ids]
```

(the init tokens is a bit janky above, it would be cleaner if we didn't do the deepcopy above since then it is a list of lists).

Comment (naveen-corpusant) [reply src/transformers/models/whisper/generation_whisper.py:1194]: you can validate the current failure with the test of multiple languages, but passing `None` as the language

Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1194]: True! But it's becoming a bit convoluted then. It's probably cleaner to revert 0b424f7 so that `init_tokens` and `language` are always of length `batch_size`. @ylacombe?

**Comment by ylacombe** [comment]
> @ylacombe I ran the slow tests in test_modeling_whisper. I get some failing tests, but the same ones fail on main as well, with the same output:

As long as this PR doesn't introduce new failing tests, this should be okay!

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (ylacombe) [reply src/transformers/models/whisper/generation_whisper.py:1194]: Not sure to follow the issue here, could you expand on this ?

Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1194]: @ylacombe 0b424f7 was based on [your suggestion](https://github.com/huggingface/transformers/pull/29688#discussion_r1528385312), but it turns out it broke language detection. So maybe we can revert that unless you think it would introduce significant overhead.

Another option would be to do something like this:
```python
        if isinstance(language, (list, tuple)):
            # ...
        elif language is None:
            # Language will be detected for each item in batch
            language = [None] * batch_size
        else:
            language = [language]  # Use a length-1 list now, broadcast later

        # Separate init_tokens for each language
        init_tokens = [copy.deepcopy(init_tokens) for _ in language]
```

File: src/transformers/models/whisper/generation_whisper.py [hunk 4]
```diff
@@ -1067,7 +1068,10 @@ def _set_language_and_task(language, task, is_multilingual, generation_config):
                     "to `generate`. Either set the language using the `forced_decoder_ids` in the model config, "
                     "or update the generation config as per the instructions https://github.com/huggingface/transformers/issues/25084#issuecomment-1664398224"
                 )
-            language = language.lower()
+            if isinstance(language, str):
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1071]: I would be in favour of moving this normalisation of the `language` argument to the `_retrieve_init_tokens` method. It can be executed just before we map the `language` string argument to the langauge id token: https://github.com/huggingface/transformers/blob/9b0a8ea7d1d6226b76cfdc645ce65e21157e2b50/src/transformers/models/whisper/generation_whisper.py#L1138

File: src/transformers/models/whisper/generation_whisper.py [hunk 5]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
                     f"You are using token ids in `forced_decoder_ids` that do not seem to correctly follow the prompt pattern of Whisper. Make sure that {forced_decoder_ids} has an entry for all indices >= 1 and < {forced_decoder_ids[0][0]}.",
                 )
 
-        # from v4.39 the forced decoder ids are always None in favour of decoder input ids
-        generation_config.forced_decoder_ids = None
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1135]: Is this an intentional deletion? Note that we're moving away from `forced_decoder_ids` in favour of the `language`/`task` arguments

File: src/transformers/models/whisper/generation_whisper.py [hunk 6]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
-                is_language_code = len(language) == 2
-                raise ValueError(
-                    f"Unsupported language: {language}. Language should be one of:"
-                    f" {list(TO_LANGUAGE_CODE.values()) if is_language_code else list(TO_LANGUAGE_CODE.keys())}."
+
+        # Make sure language is a list of strings of the correct length
+        if isinstance(language, (list, tuple)):
+            if any(l is None for l in language):
+                raise TypeError(
+                    "Expected `language` to be `None`, a single string or a list of strings. Got a list containing `None`"
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1144]:
```suggestion
                    "Expected `language` to be `None`, a single string (e.g. `'en'`), or a list of strings with length equal to the batch size (e.g. `('en', 'fr')` for a batch size of 2). Got a list containing `None`."
```

File: src/transformers/models/whisper/generation_whisper.py [hunk 7]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
                 )
-            if language_token not in generation_config.lang_to_id:
+            if len(language) != batch_size:
                 raise ValueError(
-                    f"{language_token} is not supported by this specific model as it is not in the `generation_config.lang_to_id`."
-                    "(You should just add it to the generation config)"
+                    "When passing a list of languages, the length of the list must match the batch size. "
+                    f"Expected length of {batch_size}, but got {len(language)}"
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1149]: (nit)
```suggestion
                    f"Expected length of {batch_size}, but got {len(language)} languages."
```

File: src/transformers/models/whisper/generation_whisper.py [hunk 8]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
                 )
+        else:
+            language = [language]  # Use a length-1 list now, broadcast later
+
+        # Separate init_tokens for each language
+        init_tokens = [copy.deepcopy(init_tokens) for _ in language]
+
+        # from v4.39 the forced decoder ids are always None in favour of decoder input ids
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1157]: Ah ok it's added back here. Could we move it back up so it sits next to the forced decoder id logic?

File: src/transformers/models/whisper/generation_whisper.py [hunk 9]
```diff
@@ -1131,33 +1135,52 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
-                    f"{language_token} is not supported by this specific model as it is not in the `generation_config.lang_to_id`."
-                    "(You should just add it to the generation config)"
+                    "When passing a list of languages, the length of the list must match the batch size. "
+                    f"Expected length of {batch_size}, but got {len(language)}"
                 )
+        else:
+            language = [language]  # Use a length-1 list now, broadcast later
+
+        # Separate init_tokens for each language
+        init_tokens = [copy.deepcopy(init_tokens) for _ in language]
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1155]: Is there a reason we need a `deepcopy` here?

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1194]: The rest code assumes that we're using the same init tokens across the entire batch, but in the proposed changes we're looping over each batch item and computing the init tokens for each. This is redundant since the init tokens will be the same for each element of the batch. In practice, we should only need to compute them once!

Instead of doing this looping, can we keep the existing code and compute the init tokens just once? We can then copy them for each element in the batch, in a similar way to how you did previously:

```python
        # Separate init_tokens for each language
        init_tokens = [copy.deepcopy(init_tokens) for _ in language]
```

File: tests/models/whisper/test_modeling_whisper.py [hunk 10]
```diff
@@ -1748,29 +1757,44 @@ def test_large_generation_multilingual(self):
...
             input_features, do_sample=False, max_length=20, language="<|en|>", task="transcribe"
         )
         transcript = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
 
-        EXPECTED_TRANSCRIPT = " Kimura-san called me."
-        self.assertEqual(transcript, EXPECTED_TRANSCRIPT)
+        EXPECTED_TRANSCRIPT_EN = " Kimura-san called me."
+        self.assertEqual(transcript, EXPECTED_TRANSCRIPT_EN)
 
+        # Both languages in the same batch
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:1778]: Nice test! Would you mind defining a new test function for this test and moving the new checks into here? (e.g. `test_large_generation_batched_languages`) Keeping the tests small and modular helps identify which parts of the code are potentially broken with new PRs

**Review by sanchit-gandhi — COMMENTED** [review]
Thank you for your contribution @cifkao! The PR looks nearly ready to go, with just some small refactor suggestions below. Since this PR touches some core generation code in Whisper, could you please run the slow tests for the model and confirm that they pass before merging? No worries if you don't have the compute to do this, we can also run it on a machine and copy back the results (although it'll certainly be faster for you to iterate if you have a machine you can use!). Also cc @kamilakesbi for visibility

File: src/transformers/models/whisper/generation_whisper.py (hunk 4 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1071]: @sanchit-gandhi But this is already the `_retrieve_init_tokens` function. But yes, I'll move it here to normalize the list of languages once it's already been ensured that it's a list.

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1194]: This is already happening. `init_tokens` is of length 1, except if a list of languages is passed.

File: src/transformers/models/whisper/generation_whisper.py (hunk 9 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1155]: Not really, will change it to `copy` then.

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1194]: Actually @naveen-corpusant, I don't think you're right on your second point. That later check is not checking the detected language, it's there to check if language was passed to `generate()` and if so, set the task to transcription. I'll introduce a variable `languages` in order to keep the original `language` parameter intact.

File: src/transformers/models/whisper/generation_whisper.py [hunk 11]
```diff
@@ -1135,81 +1157,83 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
+            languages = [None] * batch_size
+        else:
+            languages = [language]  # Use a length-1 list now, broadcast later
 
-            lang_id = generation_config.lang_to_id[language_token]
+        # Separate init_tokens for each language
+        init_tokens = [copy.copy(init_tokens) for _ in languages]
 
-            # if language is defined it'll overwrite language ids that might have already been defined via the generation_config
-            replace_or_add(init_tokens, lang_id, generation_config.lang_to_id.values())
```
Comment (cifkao) [inline src/transformers/models/whisper/generation_whisper.py:1160]: I'm not sure why this function was used here. The language token is always `init_tokens[1]`, so it's doing the same thing as this bit in the `elif` branch (only slower): https://github.com/huggingface/transformers/blob/56b64bf1a51e29046bb3f8ca15839ff4d6a92c74/src/transformers/models/whisper/generation_whisper.py#L1177-L1181
So I unified it.

File: tests/models/whisper/test_modeling_whisper.py (hunk 10 above)
Comment (cifkao) [reply tests/models/whisper/test_modeling_whisper.py]: Done!

**Review by cifkao — COMMENTED** [review]
@ylacombe @sanchit-gandhi The code was quite hard to follow, so I refactored it a bit.

I re-ran the slow tests and got the same output – a few failing tests, but last time I checked, they were failing on `main` as well:
```
========================================================================================================================== short test summary info ===========================================================================================================================
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_multi_batch - assert [' While Porashaggy sits there, a cooing dove. He has gone, and gone for good," answered Polychrom, who had managed to squeeze into the room beside the dragon, and had witnessed the occurrences with much interest. I have remained a prisoner only because I wi...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_multi_batch_prev_cond - assert [" While poor shaggy sits there, a cooling dove. He has gone and gone for good, answered polychrome, who had managed to squeeze into the room beside the dragon, and had witnessed the occurrences with much interest. I have remained a prisoner only because I w...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_prompt_ids - AssertionError: assert 'quilter' in 'because you were sleeping instead of conquering the lovely rose princess has become a fiddle without a bow while poor shaggy sits there a cooing dove'
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_single_batch - assert [" Because you were sleeping instead of conquering, the lovely rose princess has become a fiddle without a bow, all poor ashaggy sits there, accooing dove. He has gone and gone for good. Antered polychrome who had managed to squeeze into the room beside the ...
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_whisper_longform_single_batch_prev_cond - assert [" Because you were sleeping instead of conquering, the lovely rose princess has become a fiddle without a bow, all poor ashaggy sits there, accooing dove. He has gone and gone for good. Antered polychrome who had managed to squeeze into the room beside the ...
==================================================================================================== 5 failed, 293 passed, 106 skipped, 121 warnings in 259.68s (0:04:19) ====================================================================================================
```

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (cifkao) [reply src/transformers/models/whisper/generation_whisper.py:1196]: I went with my suggestion above, so `language` will be a length-1 list if a single language was passed, otherwise it will be a length-`batch_size` list of either languages or `None`s (if the languages need to be detected).

File: src/transformers/models/whisper/generation_whisper.py [hunk 12]
```diff
@@ -1135,81 +1157,83 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
...
+            # append or replace lang_ids to init_tokens
+            for i in range(len(init_tokens)):
+                if len(init_tokens[i]) > 1:
+                    init_tokens[i][1] = lang_ids[i]
+                else:
+                    init_tokens[i].append(lang_ids[i])
+        del languages
+
+        # Update init_tokens with task
+        for i in range(len(init_tokens)):
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1204]: This looks much cleaner and I understand the motivation for looping over the `init_tokens`!

**Review by sanchit-gandhi — APPROVED** [review]
Very clean PR - thanks for iterating @cifkao!

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/generation_whisper.py:1196]: The current version of the code makes it clear why we need to do this - happy with how it is! https://github.com/huggingface/transformers/pull/29688#discussion_r1559490951

**Comment by sanchit-gandhi** [comment]
Failing slow tests will be addressed by #30152

**Review by kamilakesbi — APPROVED** [review]
Hi @cifkao, 

Thanks for working on this! 
LGTM

**Comment by ArthurZucker** [comment]
Reviewing in a bit to finish it! Would be nice to rebase as it's been a while! 🤗

**Review by ArthurZucker — APPROVED** [review]
Looks good to me! 
This is a nice feature, thanks for adding support for batches!

**Comment by ArthurZucker** [comment]
can you merge with main to make sure all the tests pass? 🤗

**Comment by cifkao** [comment]
@ArthurZucker Done!

**Comment by ArthurZucker** [comment]
Congrats for this PR! 🤗

**Comment by AvivSham** [comment]
Hi @ArthurZucker @sanchit-gandhi @cifkao,
Thank you for this PR much needed. 
When calling generate:
```python 
model.generate(
                batch.input_features,
                language=batch.language,
            )
```

We observe the following warning:
```
You have passed language=['en', 'fr'], but also have set `forced_decoder_ids` to [[1, None], [2, 50359]] which creates a conflict. `forced_decoder_ids` will be ignored in favor of language=['en', 'fr'].
The attention mask is not set and cannot be inferred from input because pad token is same as eos token.As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
```

Can it be ignored? Since we are in generate mode, the pad token does not affect the process.
